src/hp/derp/server.rs

```diff
@@ -368,26 +305,192 @@ pub(crate) struct ServerInfo {
 use super::client_conn::ClientBuilder;
 // TODO: need a types module
 use super::client_conn::Packet;
+use super::client_conn::PeerConnState;
+
+use tokio::sync::mpsc;
+use tokio_util::sync::CancellationToken;
+
+pub(crate) struct ServerActor<C, R, W, P>
+where
+    C: Conn,
+    R: AsyncRead + Unpin + Send + Sync + 'static,
+    W: AsyncWrite + Unpin + Send + Sync + 'static,
+    P: PacketForwarder,
+{
+    key: PublicKey,
+    receiver: mpsc::Receiver<ServerMessage<C, R, W, P>>,
+    /// All clients connected to this server
+    clients: Clients<C>,
+    /// Representation of the mesh network. Keys that are associated with `None` are strictly local
+    /// clients.
+    client_mesh: HashMap<PublicKey, Option<P>>,
+    /// Mesh clients that need to be appraised on the state of the network
+    watchers: HashSet<PublicKey>,
+}
+
+impl<C, R, W, P> ServerActor<C, R, W, P>
+where
+    C: Conn,
+    R: AsyncRead + Unpin + Send + Sync + 'static,
+    W: AsyncWrite + Unpin + Send + Sync + 'static,
+    P: PacketForwarder,
+{
+    pub(crate) fn new(key: PublicKey, receiver: mpsc::Receiver<ServerMessage<C, R, W, P>>) -> Self {
+        Self {
+            key,
+            receiver,
+            clients: Clients::new(),
+            client_mesh: HashMap::default(),
+            watchers: HashSet::default(),
+        }
+    }
+
+    pub(crate) async fn run(mut self, done: CancellationToken) -> Result<()> {
+        loop {
+            tokio::select! {
+                biased;
+                _ = done.cancelled() => {
```

Comment:
This is a hard shutdown. It's good to have it, but if you also want a soft shutdown you can just make another normal message. That way you process all incoming msgs up to the shutdown msg.